openshift-kube-apiserver/admission/storage/performantsecuritypolicy/admission.go

```diff
@@ -0,0 +1,167 @@
+	kapi "k8s.io/kubernetes/pkg/apis/core"
+)
+
+const (
+	PluginName               = "storage.openshift.io/PerformantSecurityPolicy"
+	fsGroupChangePolicyLabel = "storage.openshift.io/fsgroup-change-policy"
+	selinuxChangePolicyLabel = "storage.openshift.io/selinux-change-policy"
+)
+
+var (
+	_ = initializer.WantsExternalKubeInformerFactory(&performantSecurityPolicy{})
+	_ = admission.MutationInterface(&performantSecurityPolicy{})
+	_ = initializer.WantsFeatures(&performantSecurityPolicy{})
+)
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(config io.Reader) (admission.Interface, error) {
+			return &performantSecurityPolicy{
+				Handler: admission.NewHandler(admission.Create),
+			}, nil
+		})
+}
+
+// performantSecurityPolicy checks and applies if a default FSGroupChangePolicy and SELinuxChangePolicy
+// should be applied to the pod.
+type performantSecurityPolicy struct {
+	*admission.Handler
+	storagePerformantSecurityPolicyFeatureEnabled bool
+	nsLister                                      corev1listers.NamespaceLister
+}
+
+// SetExternalKubeInformerFactory registers an informer
+func (c *performantSecurityPolicy) SetExternalKubeInformerFactory(kubeInformers informers.SharedInformerFactory) {
+	c.nsLister = kubeInformers.Core().V1().Namespaces().Lister()
+	c.SetReadyFunc(func() bool {
+		return kubeInformers.Core().V1().Namespaces().Informer().HasSynced()
+	})
+}
+
+func (c *performantSecurityPolicy) InspectFeatureGates(featureGates featuregate.FeatureGate) {
+	c.storagePerformantSecurityPolicyFeatureEnabled = featureGates.Enabled(featuregate.Feature(openshiftfeatures.FeatureGateStoragePerformantSecurityPolicy))
+}
+
+func (c *performantSecurityPolicy) ValidateInitialization() error {
+	if c.nsLister == nil {
+		return fmt.Errorf("%s plugin needs a namespace lister", PluginName)
+	}
+	return nil
+}
+
+func (c *performantSecurityPolicy) Admit(ctx context.Context, attributes admission.Attributes, _ admission.ObjectInterfaces) error {
+	if !c.storagePerformantSecurityPolicyFeatureEnabled {
+		return nil
+	}
+
+	if !c.WaitForReady() {
+		return admission.NewForbidden(attributes, fmt.Errorf("not yet ready to handle request"))
+	}
+
+	if attributes.GetResource().GroupResource() != kapi.Resource("pods") ||
+		len(attributes.GetSubresource()) > 0 {
+		return nil
+	}
+
+	pod, ok := attributes.GetObject().(*kapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attributes, fmt.Errorf("unexpected object: %#v", attributes.GetObject()))
+	}
+
+	ns, err := c.nsLister.Get(pod.Namespace)
+	if err != nil {
+		return fmt.Errorf("error listing pod namespace: %v", err)
+	}
+
+	currentFSGroupChangePolicy := extractCurrentFSGroupChangePolicy(pod)
+	if currentFSGroupChangePolicy == nil {
+		currentFSGroupChangePolicy = getDefaultFSGroupChangePolicy(ns)
+		if currentFSGroupChangePolicy != nil {
+			klog.V(3).Infof("Setting default FSGroupChangePolicy %s for pod %s/%s", *currentFSGroupChangePolicy, pod.Namespace, pod.Name)
```

Comment:
Consider adding this information to the audit logs if you'd like to examine what happened to this resource.

Take a look at https://github.com/kubernetes/kubernetes/blob/27e16751538cb316f0c8b8f2384a9968df066baa/plugin/pkg/admission/security/podsecurity/admission.go#L200  as an example from an upstream plugin.

Reply:
Is it okay if we merge this PR (unless your comment is a blocker) and put this feature through some testing in CI before we start adding more changes? 

Please note - eventually we want move this feature to use MAP, whenever MAP becomes GA in OCP. 

Audit logs is probably nice to have here - we did not document this as a requirement in - https://github.com/gnufied/enhancements-1/blob/3a3eba791fad480a059aaa5bd07331a4efe36b80/enhancements/storage/storage-performant-security-policy.md